src/sancus-step/sancus_step.c

```diff
@@ -1,11 +1,96 @@
 #include "sancus_step.h"
 #include "timer.h"
+#include <stdio.h>
+#include <sancus/sm_support.h>
+#include "sm_io.h"
 
-void*    __ss_isr_sp = &__ss_isr_stack[ISR_STACK_SIZE-1];
+void sancus_step_print_latency(void)
+{
+    int latency = sancus_step_get_latency();
+    printf("latency: %d\n", latency);
+}
+
+int sancus_step_get_latency(void)
+{
+    return __ss_isr_tar_entry - sm_exit_latency;
+}
+
+int ss_dbg_entry_delay = 0;
+int ss_dbg_measuring_reti_latency = 0;
+int isr_reti_latency = 0;
+int sm_exit_latency = 0;
+int isr_interrupted_sm = 0;
+
+/*
+ * Determines isr_reti_latency (reti into interrupted module)
+ * and sm_exit_latency (from interrupt to isr)
+ * and starts timer, so that nemesis can be executed
+ */
+void sancus_step_start(void)
+{
+    sancus_step_init();
+    __asm__("dint\n\t");
+    timer_irq(INIT_LATENCY);
+}
 
-void print_latency(void)
+/*
+ * Determines isr_reti_latency (reti into interrupted moduled)
+ * and sm_exit_latency (from interrupt to isr)
+ */
+void sancus_step_init(void)
+{
+    sancus_enable(&ssdbg);
+    
+    ss_dbg_measuring_reti_latency = 1;
+    
+    timer_tsc_start();
+    ss_dbg_get_info();
+    
+    timer_irq(ss_dbg_entry_delay);
+    int isr_reti_latency_t = ss_dbg_get_info();
+    /* amount of cycles in the reti logic = measured delay
+                                  - record delay
+                                  - duration of reti instruction
+                                  + 1 (because we count amount of cycles)
+    */
+    isr_reti_latency = isr_reti_latency_t
+                        - JMP_LENGTH
+                        - RETI_LENGTH
+                        + 1;
+    
+    timer_irq(ss_dbg_entry_delay + EXTRA_DELAY);
+    ss_dbg_get_info();
+    sm_exit_latency = __ss_isr_tar_entry - 3;
+    printf("%d, %d, %d\n", isr_reti_latency, sm_exit_latency, ss_dbg_entry_delay);
+    
+    ss_dbg_measuring_reti_latency = 0;
+}
+
+void sancus_step_end(void)
 {
     TACTL = TACTL_DISABLE;
-    int latency = __ss_isr_tar_entry - HW_IRQ_LATENCY - 1;
-    printf("latency: %d\n", latency);
+}
+
+DECLARE_SM(ssdbg, 0x1234);
+
+__attribute__((naked))
+int SM_ENTRY(ssdbg) ss_dbg_get_info(void)
+{
+    /*
```

Comment:
let's also remove this commented code for clarity and just add comments between the asm where the IRQ is supposed to arrive?